### DISCUSSION TITLE
Xdebug slows down processing

### DISCUSSION THREAD

**Discussion by iGrog** [roadrunner-server/roadrunner#1457] [Q&A]
RoadRunner: 2.12.2
Spiral Framework: 3.5

Running app on Windows under WLS2 in Docker

rr.yaml
```
version: '2.7'

rpc:
  listen: tcp://127.0.0.1:6001

server:
  command: "php app.php"
  relay: pipes
  env:
    PHP_IDE_CONFIG: "serverName=API"
    XDEBUG_SESSION: 1
# HTTP plugin settings
http:
  address: 0.0.0.0:8080
  middleware: [ "gzip", "static", "headers" ]
  static:
    dir: "public"
    forbid: [ ".php", ".htaccess" ]
  pool:
    debug: true
    num_workers: 1
    supervisor:
      max_worker_memory: 256
```

Dockerfile:
```
ARG PHP_IMAGE=8.1-cli-alpine
ARG ROAD_RUNNER_IMAGE=2.12.2

#Download rr binary
FROM spiralscout/roadrunner:$ROAD_RUNNER_IMAGE as builder

FROM php:$PHP_IMAGE

#... install common php extensions

COPY --from=composer:2.2.7 /usr/bin/composer /usr/bin/composer

WORKDIR /var/www

COPY --from=builder /usr/bin/rr /var/server/
COPY ./.rr.yaml /var/server/.rr.yaml
COPY ./wait-for-it.sh /var/server/
RUN ["chmod", "+x", "/var/server/wait-for-it.sh"]

# app dramatically slows down if uncomment this line and rebuild container
#RUN pecl install xdebug && docker-php-ext-enable xdebug 

COPY ./docker/php/php.ini /usr/local/etc/php
COPY ./docker/php/conf.d /usr/local/etc/php/conf.d

CMD ["/var/server/rr", "serve", "-d"]
```

WebController
```
    #[Route(route: '/', name: 'index', methods: ['GET'])]
    public function index(): string
    {
        return 'Hello world';
    }
```
Using `Route` attributes.
With commented xdebug running time about 166ms (rr debug: true)
![image](https://user-images.githubusercontent.com/842588/218254342-6d0ca7ad-9f25-4d6b-92ab-9d5e48578fb5.png)

Profiler buggerator:
![image](https://user-images.githubusercontent.com/842588/218254419-e6f31311-1d1e-442f-8ddf-b85d48130fff.png)
Json log here: https://www.npoint.io/docs/472558ad8c862954dda2

If install x-debug: `RUN pecl install xdebug && docker-php-ext-enable xdebug`
and run the same action, execution time is slower around 1.5s
![image](https://user-images.githubusercontent.com/842588/218255002-a355c14f-5a82-4847-95f9-67e4c17f13c9.png)

![image](https://user-images.githubusercontent.com/842588/218255016-09d0e609-1e56-4f95-bed9-16b773cd0f4f.png)
Full json here: https://www.npoint.io/docs/0451138e220f1270b7d5
in RR Log: `"elapsed":"1.60626074s"`
```
{"level":"DEBUG","ts":"2023-02-11T11:13:50.742Z","logger":"server      ","msg":"worker is allocated","pid":33,"internal_event_name":"EventWorkerConstruct"}
{"level":"DEBUG","ts":"2023-02-11T11:13:50.942Z","logger":"server      ","msg":"sending stop request to the worker","pid":33}
{"level":"INFO","ts":"2023-02-11T11:13:50.949Z","logger":"http        ","msg":"http log","status":200,"method":"GET","URI":"/","remote_address":"192.168.48.1:37076","read_bytes":0,"write_bytes":11,"start":"2023-02-11T11:13:49.342Z","elapsed":"1.60626074s"}

```

Question: what slows down for an extra 1.5 seconds?
What to do?

**Comment by rustatian** [comment]
Hey @iGrog 👋🏻 

What is your XDebug configuration (php.ini/xdebug.conf)?

**Reply by iGrog** [reply]
xdebug.mode=debug,develop
xdebug.client_host=host.docker.internal
xdebug.client_port=9000
xdebug.var_display_max_depth=10

**Reply by rustatian** [reply]
ping: @butschster, @wolfy-j

**Comment by roxblnfk** [comment]
Do you have the same processing time with RoadRunner `debug: false` mode (in the second request on the same worker)?

**Reply by iGrog** [reply]
No, with `debug:false` it is much faster.  
I cloned your `app` repository to eliminate my own code.
Controller looks like:
```
    public function index(): string
    {
        return 'Hello world';
    }
```    

First 4 requests with RR `debug: true`
Next 4 requests with RR `debug:false`

XDebug installed in both runs. But it doesn't work when `debug:false` (no stops at breakpoints)
![image](https://user-images.githubusercontent.com/842588/220460839-0f056bbe-4a5c-4955-afff-f608432e98b7.png)
Both runs used  `TOKENIZER_CACHE_TARGETS=true`

**Reply by rustatian** [reply]
This is expected behavior. When you use a `debug: true`, RR won't allocate workers initially (your configuration for that will be ignored). After you send a request, RR starts a worker and kills it after sending the response back. With this behavior, you'll always have the latest code for your worker when you initiate a request. So, debug - no caching, no warm-up for the PHP opcache. Request - worker up - response - worker down.

**Reply by iGrog** [reply]
> This is expected behavior

What exactly is expected behaviour?
Same code on `Slim`+`php-fpm`+`nginx` with `xdebug` turned on executes much much faster.

Are there any ways to speed up xdebug to decrease response time in `debug: true` mode?

**Reply by rustatian** [reply]
> What exactly is expected behavior?

Slowness is expected. What is the execution time with `xdebug` without the `debug: true` ?

**Reply by iGrog** [reply]
Around 150ms. But breakpoints don't work in that case.

**Reply by iGrog** [reply]
https://github.com/spiral/framework/issues/894 I found that GroupRegistry::registerRoutes very slow with xdebug.

**Comment by tntrex** [comment]
Hello. I'm using roadrunner in docker in wsl2 too. You can try to configure/disable firewall or set [`xdebug.connect_timeout_ms`](https://xdebug.org/docs/all_settings#connect_timeout_ms) to very small number, like 5ms.
